### PR TITLE
image/preseed, cmd/snap-preseed: create and sign preseed assertion

### PR DESCRIPTION
**Pull request by stolowski**
Create preseed assertion and sign it with the key provided via preseed options (or cli).

### CODE REVIEW

File: image/preseed/preseed_uc20_test.go [hunk 1]
```diff
@@ -27,13 +27,91 @@ import (
...
+				"run": {{
+					Path: "/some/path/foo.snap",
+					SideInfo: &snap.SideInfo{
+						RealName: "foo"},
+				}}},
+		}, nil
+	})
+	defer restoreSeedOpen()
+
+	restoreSaveAssertion := preseed.MockSaveAssertion(func(*asserts.Database, asserts.Assertion, *asserts.Model) error {
```
Comment (pedronis) [inline image/preseed/preseed_uc20_test.go:103]: this is mocking too much here,  this way we don't need to deal with the account-key that corresponds to the key we create below, but it's important that the real code saves that account-key as I mentioned elsewhere

File: image/image_linux.go [hunk 2]
```diff
@@ -264,6 +267,22 @@ func makeLabel(now time.Time) string {
...
+func saveAssertion(db *asserts.Database, as asserts.Assertion, model *asserts.Model) error {
+	tsto, err := newToolingStoreFromModel(model, "")
+	if err != nil {
+		return err
+	}
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		return tsto.AssertionFetcher(db, save)
+	}
+
+	f := seedwriter.MakeRefAssertsFetcher(newFetcher)
```
Comment (pedronis) [inline image/image_linux.go:279]: to do the right thing with respect to account-key we will need to get hold of f to look at f.Refs() I think

**Review by pedronis — COMMENTED** [review]
thank you, did a pass, the tests are not testing enough, the code is not dealing with saving the account-key that corresponds to the key used to sign the preseed assertions. Also the SaveAssertion hook is awkward

File: cmd/snap/cmd_prepare_image.go [hunk 3]
```diff
@@ -132,6 +138,9 @@ func (x *cmdPrepareImage) Execute(args []string) error {
 	opts.PrepareDir = x.Positional.TargetDir
 	opts.Classic = x.Classic
 
+	opts.Preseed = x.Preseed
+	opts.PreseedSignKey = x.PreseedSignKey
```
Comment (pedronis) [inline cmd/snap/cmd_prepare_image.go:142]: should this error if Preseed is false but PreseedSignKey is set? or should it set Preseed implicitly? we need to do one of those things I think

File: image/image_test.go [hunk 4]
```diff
@@ -3270,9 +3270,10 @@ func (s *imageSuite) TestPrepareWithUC20Preseed(c *C) {
 	defer restoreSetupSeed()
 
 	var preseedCalled bool
-	restorePreseedCore20 := image.MockPreseedCore20(func(dir string) error {
+	restorePreseedCore20 := image.MockPreseedCore20(func(dir, key string) error {
 		preseedCalled = true
 		c.Assert(dir, Equals, "/a/dir")
+		c.Assert(key, Equals, "")
```
Comment (pedronis) [inline image/image_test.go:3276]: shouldn't the test be changed to pass a key so that we see it arrive here?

File: image/preseed/preseed.go [hunk 5]
```diff
@@ -39,9 +54,153 @@ type preseedOpts struct {
...
+	headers := map[string]interface{}{
+		"type":              "preseed",
+		"authority-id":      model.AuthorityID(),
+		"series":            "16",
+		"brand-id":          model.BrandID(),
+		"model":             model.Model(),
+		"system-label":      opts.SystemLabel,
+		"artifact-sha3-384": base64Digest,
+		"timestamp":         time.Now().UTC().Format(time.RFC3339),
+		"revision":          "1",
```
Comment (pedronis) [inline image/preseed/preseed.go:164]: we don't need/want this. We want this to have the default revision of 0

File: image/preseed/preseed.go [hunk 6]
```diff
@@ -39,9 +54,153 @@ type preseedOpts struct {
...
+		"snaps":             snaps,
+	}
+
+	signedAssert, err := adb.Sign(asserts.PreseedType, headers, nil, privKey.PublicKey().ID())
+	if err != nil {
+		return fmt.Errorf("cannot sign preseed asertion: %v", err)
```
Comment (pedronis) [inline image/preseed/preseed.go:170]: typo: assertion

File: image/preseed/preseed.go [hunk 7]
```diff
@@ -39,9 +54,153 @@ type preseedOpts struct {
...
+	if err != nil {
+		return err
+	}
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		return tsto.AssertionFetcher(adb, save)
+	}
+
+	f := seedwriter.MakeRefAssertsFetcher(newFetcher)
+	if err := f.Save(signedAssert); err != nil {
+		return fmt.Errorf("cannot save assertion: %v", err)
```
Comment (pedronis) [inline image/preseed/preseed.go:183]: s/save/fetch/ as if there will be error they will be about the assertions we don't have

File: image/preseed/preseed.go [hunk 8]
```diff
@@ -39,9 +54,153 @@ type preseedOpts struct {
...
+
+	enc := asserts.NewEncoder(serialized)
+	for _, aref := range f.Refs() {
+		if aref.Type == asserts.PreseedType || aref.Type == asserts.AccountKeyType {
+			as, err := aref.Resolve(adb.Find)
+			if err != nil {
+				return fmt.Errorf("internal error: %v", err)
+			}
+			if err := enc.Encode(as); err != nil {
+				return fmt.Errorf("cannot save assertion %s: %v", aref, err)
```
Comment (pedronis) [inline image/preseed/preseed.go:200]: s/save/write/

**Review by pedronis — COMMENTED** [review]
looking good, did a pass, mostly small remarks

File: image/preseed/preseed_uc20_test.go [hunk 9]
```diff
@@ -20,20 +20,139 @@
...
+			SnapsForMode: map[string][]*seed.Snap{
+				"run": {{
+					Path: "/some/path/foo.snap",
+					SideInfo: &snap.SideInfo{
+						RealName: "foo"},
+				}}},
+			loadAssertions: func(db asserts.RODatabase, commitTo func(*asserts.Batch) error) error {
+				// XXX: this doesn't have effect on the serialized account-keys
+				// batch := asserts.NewBatch(nil)
+				// c.Assert(batch.Add(ts.StoreSigning.TrustedKey), IsNil)
```
[the remark this answers is not in the document]
Comment (stolowski) [reply image/preseed/preseed_uc20_test.go]: You're right, that fixed it, thanks!

File: image/preseed/preseed.go [hunk 10]
```diff
@@ -161,13 +161,13 @@ func writePreseedAssertion(opts *preseedOpts, artifactDigest []byte) error {
 		"system-label":      opts.SystemLabel,
 		"artifact-sha3-384": base64Digest,
 		"timestamp":         time.Now().UTC().Format(time.RFC3339),
-		"revision":          "1",
+		"revision":          "0",
```
Comment (pedronis) [inline image/preseed/preseed.go:164]: nitpick: you can leave it out, and it will be assumed as it's the default

**Review by pedronis — APPROVED** [review]
thank you, small nitpick

File: image/preseed/preseed.go [hunk 11]
```diff
@@ -39,9 +54,153 @@ type preseedOpts struct {
...
+	}
+
+	key := opts.PreseedSignKey
+	if key == "" {
+		key = `default`
+	}
+	privKey, err := keypairMgr.GetByName(key)
+	if err != nil {
+		// TRANSLATORS: %q is the key name, %v the error message
+		return fmt.Errorf(i18n.G("cannot use %q key: %v"), key, err)
```
Comment (bboozzoo) [inline image/preseed/preseed.go:92]: This error uses translations, but other errors do not. Most likely the error from GetByName() aren't translated either.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: image/preseed/preseed.go (hunk 11 above)
Comment (stolowski) [reply image/preseed/preseed.go:92]: Indeed. For the time being I'll leave it as is,`cmd_sign.go` (that I based this on) has the same property.

**Pull request merged**